File: SIGAI/main/Mode.swift
//
//  Mode.swift
//  SIGAI
//
//  Created by Ainal syazwan Itamta on 22/04/2025.
//

import SwiftUI
import AVFoundation

extension ContentView {
    func playSound(success: Bool) {
        if success {
            let soundID: SystemSoundID = 1335 // Apple Pay Success-like sound
            AudioServicesPlaySystemSound(soundID)
        } else {
            let soundID: SystemSoundID = 1006 // Strong error thunk
            AudioServicesPlaySystemSound(soundID)
            AudioServicesPlaySystemSound(kSystemSoundID_Vibrate) // Vibrate when wrong
        }
    }
    func ModeView(geometry: GeometryProxy) -> some View {
        ZStack {
            LinearGradient (
                gradient: Gradient(colors: isDarkMode ? [
                    Color.black,
                    Color(red: 230/255, green: 230/255, blue: 230/255),
                    Color.white,
                    Color(red: 245/255, green: 245/255, blue: 245/255),
                    Color.white,
                    Color(red: 250/255, green: 250/255, blue: 250/255),
                    Color.white
                ] : [
                    Color(red: 235/255, green: 250/255, blue: 255/255),
                    Color.white,
                    Color(red: 220/255, green: 245/255, blue: 255/255), // Very light blue
                    Color.white,
                    Color(red: 255/255, green: 220/255, blue: 230/255)  // Softer pink
                ]),
                startPoint: .top,
                endPoint: .bottom
            )
            .edgesIgnoringSafeArea(.all)
            
            VStack {
                sigaiProgressBar(appLanguage: appLanguage)
                    .frame(maxWidth: .infinity)
                VStack {
                    ZStack {
                        RoundedRectangle(cornerRadius: 16)
                            .fill(
                                LinearGradient(
                                    gradient: Gradient(colors: isDarkMode ? [
                                        Color(red: 50/255, green: 50/255, blue: 100/255), // Dark Blue
                                        Color(red: 80/255, green: 80/255, blue: 150/255)  // Slightly lighter Dark Blue
                                    ] : [
                                        Color(red: 1.0, green: 0.8, blue: 0.9), // Soft Pink
                                        Color(red: 210/255, green: 240/255, blue: 255/255) // Light Blue
                                    ]),
                                    startPoint: .top,
                                    endPoint: .bottom
                                )
                            )
                            .shadow(color: Color.black.opacity(0.1), radius: 6, x: 0, y: 3)
                            .frame(maxWidth: .infinity)
                        VStack {
                            if mode == .multiplication {
                                Text(appLanguage == "ms" ? "Hasil Darab: \(question)" : "Multiplication Result: \(question)")
                                    .font(.title2)
                                    .fontWeight(.semibold)
                                    .foregroundColor(isDarkMode ? Color.yellow.opacity(0.8) : Color.blue.opacity(0.7))
                                    .shadow(color: (isDarkMode ? Color.yellow.opacity(0.2) : Color.blue.opacity(0.2)), radius: 5, x: 0, y: 2)
                                Text("\(countIntersections())")
                                    .font(.title2)
                                    .fontWeight(.semibold)
                                    .background(Color.clear)
                                    .cornerRadius(5)
                                    .frame(width: 200, height: 40)
                                    .multilineTextAlignment(.center)
                                    .foregroundColor(isDarkMode ? Color.yellow.opacity(0.8) : Color.blue.opacity(0.7))
                                    .shadow(color: (isDarkMode ? Color.yellow.opacity(0.2) : Color.blue.opacity(0.2)), radius: 5, x: 0, y: 2)
                            } else {
                                Text(appLanguage == "ms" ? "Hasil Bahagi (\(detectedVerticalLines) garis): \(question)" : "Division Result (\(detectedVerticalLines) line/s): \(question)")
                                    .font(.title2)
                                    .fontWeight(.semibold)
                                    .foregroundColor(isDarkMode ? Color.yellow.opacity(0.8) : Color.blue.opacity(0.7))
                                    .shadow(color: (isDarkMode ? Color.yellow.opacity(0.2) : Color.blue.opacity(0.2)), radius: 5, x: 0, y: 2)
                                Text(String(format: "%.6f", countDivisions()))
                                    .font(.title2)
                                    .fontWeight(.semibold)
                                    .background(Color.clear)
                                    .cornerRadius(5)
                                    .frame(width: 300, height: 40)
                                    .multilineTextAlignment(.center)
                                    .foregroundColor(isDarkMode ? Color.yellow.opacity(0.8) : Color.blue.opacity(0.7))
                                    .shadow(color: (isDarkMode ? Color.yellow.opacity(0.2) : Color.blue.opacity(0.2)), radius: 5, x: 0, y: 2)
                            }
                            Text(appLanguage == "ms" ? "Lukis jawapan." : "Draw your answer.")
                                .font(.body)
                                .multilineTextAlignment(.center)
                                .foregroundColor(isDarkMode ? Color.white.opacity(0.8) : .secondary)
                                .shadow(color: isDarkMode ? Color.white.opacity(0.2) : Color.clear, radius: 5, x: 0, y: 2)
                                .lineLimit(nil)
                                .fixedSize(horizontal: false, vertical: true)
                                .frame(maxWidth: .infinity, alignment: .center)
                                .padding(.horizontal, 20)
                        }
                    }
                    .padding(.horizontal)
                }
                .padding(.horizontal)
                .padding(.top, geometry.safeAreaInsets.top)

                HStack {
                    Button(action: {
                        let tipsMs = [
                            "Pilih warna ikut kod: 🟣 -100, 🟡 -10, ⚫ 0, 🟢 1, 🔵 10, 🔴 100",
                            "Untuk darab, butang kunci telah dimatikan kerana tidak perlu untuk operasi darab",
                            "Untuk bahagi, lukis garisan menegak dahulu, kunci (warna kelabu), kemudian lukis garisan horizontal",
                            "Tekan 'Hantar' untuk semak jawapan!",
                            "Tekan 'Seterusnya' untuk skip ke soalan seterusnya"
                        ]
                        
                        let tipsEn = [
                            "Pick colors based on code: 🟣 -100, 🟡 -10, ⚫ 0, 🟢 1, 🔵 10, 🔴 100",
                            "For, multiplication, lock button is disabled due to not applicable",
                            "For division, draw vertical lines first, locked (grey color), then draw horizontal lines",
                            "Press 'Submit' to check the answer!",
                            "Press 'Next' to skip to the next question"
                        ]
                        
                        let tips = appLanguage == "ms" ? tipsMs : tipsEn
                        currentTip = tips[currentTipIndex]
                        currentTipIndex = (currentTipIndex + 1) % tips.count
                        showTipAlert = true
                    }) {
                        Image(systemName: "questionmark.circle.fill")
                            .font(.system(size: 32))
                            .foregroundColor(.orange)
                    }
                    
                    Button(action: {
                        if !paths.isEmpty {
                            paths.removeLast()
                            colors.removeLast()
                        }
                    }) {
                        Image(systemName: "arrow.uturn.backward.circle.fill")
                            .font(.system(size: 32))
                            .background(Color.clear)
                            .foregroundColor(.gray)
                            .cornerRadius(10)
                            .shadow(radius: 1)
                    }
                    
                    Menu {
                        Button(action: {
                            if EXPManager.level >= 20 {
                                selectedColor = .red
                            }
                        }) {
                            Label(EXPManager.level >= 20 ? "🔴 100" : (appLanguage == "ms" ? "🔴 Buka pada tahap 20" : "🔴 Unlock at Level 20"),
                                  systemImage: EXPManager.level >= 20 ? "circle.fill" : "lock.circle.fill")
                            .foregroundColor(EXPManager.level >= 20 ? .red : .gray)
                        }
                        .disabled(EXPManager.level < 20)
                        
                        Button(action: {
                            if EXPManager.level >= 10 {
                                selectedColor = .blue
                            }
                        }) {
                            Label(EXPManager.level >= 10 ? "🔵 10" : (appLanguage == "ms" ? "🔵 Buka pada tahap 10" : "🔵 Unlock at Level 10"),
                                  systemImage: EXPManager.level >= 10 ? "circle.fill" : "lock.circle.fill")
                            .foregroundColor(EXPManager.level >= 10 ? .blue : .gray)
                        }
                        .disabled(EXPManager.level < 10)
                        
                        Button(action: { selectedColor = .green }) {
                            Label("🟢 1", systemImage: "circle.fill")
                                .foregroundColor(.green)
                        }
                        
                        Button(action: { selectedColor = .gray }) {
                            Label("⚫ 0", systemImage: "circle.fill")
                                .foregroundColor(.gray)
                        }
                        
                        if mode == .division {
                            Button(action: {
                                if EXPManager.level >= 30 {
                                    selectedColor = .yellow
                                }
                            }) {
                                Label(EXPManager.level >= 30 ? "🟡 -10" : (appLanguage == "ms" ? "🟡 Buka pada tahap 30" : "🟡 Unlock at Level 30"),
                                      systemImage: EXPManager.level >= 30 ? "circle.fill" : "lock.circle.fill")
                                .foregroundColor(EXPManager.level >= 30 ? .yellow : .gray)
                            }
                            .disabled(EXPManager.level < 30)
                            
                            Button(action: {
                                if EXPManager.level >= 40 {
                                    selectedColor = .purple
                                }
                            }) {
                                Label(EXPManager.level >= 40 ? "🟣 -100" : (appLanguage == "ms" ? "🟣 Buka pada tahap 40" : "🟣 Unlock at Level 40"),
                                      systemImage: EXPManager.level >= 40 ? "circle.fill" : "lock.circle.fill")
                                .foregroundColor(EXPManager.level >= 40 ? .purple : .gray)
                            }
                            .disabled(EXPManager.level < 40)
                        }
                    } label: {
                        Image(systemName: "paintpalette.fill")
                            .font(.system(size: 32))
                            .foregroundColor(
                                selectedColor == .red ? Color.red :
                                    selectedColor == .blue ? Color.blue :
                                    selectedColor == .green ? Color.green :
                                    selectedColor == .yellow ? Color.yellow :
                                    selectedColor == .purple ? Color.purple :
                                    Color.gray
                            )
                    }
                    
                    Button(action: {
                        mode = (mode == .multiplication) ? .division : .multiplication
                        generateQuestion() // Generate a new question when mode changes
                    }) {
                        Image(systemName: mode == .multiplication ? "divide.circle.fill" : "multiply.circle.fill")
                            .font(.system(size: 32)) // Bigger size for visibility
                            .foregroundColor(.gray)
                    }
                    
                    Button(action: {
                        if mode == .multiplication {
                            if lockedVerticalLines == nil {
                                lockedVerticalLines = detectedVerticalLines
                            } else {
                                lockedVerticalLines = nil
                            }
                            
                            // Temporarily change lock color to red for 1 second
                            DispatchQueue.main.async {
                                isLocked = true
                                DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
                                    isLocked = false
                                }
                            }
                        } else {
                            if lockedVerticalLines == nil {
                                lockedVerticalLines = detectedVerticalLines
                            } else {
                                lockedVerticalLines = nil
                            }
                        }
                    }) {
                        Image(systemName: (mode == .multiplication) ? "lock.fill" : (lockedVerticalLines == nil ? "lock.open.fill" : "lock.fill"))
                            .font(.system(size: 36))
                            .foregroundColor(mode == .multiplication ? (isLocked ? .red : .gray) : (lockedVerticalLines == nil ? .blue : .gray))
                    }
                    .disabled(mode == .multiplication && isLocked)
                    
                    Button(action: {
                        paths.removeAll()
                        colors.removeAll()
                        currentPath = Path()
                        isLocked = false
                        lockedVerticalLines = nil // Unlock when clearing
                    }) {
                        Image(systemName: "eraser.fill")
                            .font(.system(size: 32))
                            .background(Color.clear)
                            .foregroundColor(.gray)
                            .cornerRadius(10)
                            .shadow(radius: 1)
                    }
                }
                .alert(isPresented: $showTipAlert) {
                    Alert(
                        title: Text("💡 Tip"),
                        message: Text(currentTip),
                        dismissButton: .default(Text("OK"))
                    )
                }
                .padding(.horizontal)

                HStack(spacing: 20) {
                    Button(action: {
                        if mode == .multiplication {
                            if countIntersections() == correctAnswer {
                                addEXP()
                                showFeedback = true
                                showCorrectOverlay = true
                                playSound(success: true)
                                DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
                                    showCorrectOverlay = false
                                    showFeedback = false
                                    generateQuestion()
                                }
                            } else {
                                showFeedback = true
                                showCorrectOverlay = false
                                playSound(success: false)
                                DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
                                    showFeedback = false
                                }
                            }
                        } else {
                            var temp_division = countDivisions() - correctAnswerB
                            if temp_division < 0 { temp_division *= -1 }
                            if temp_division < 0.00001 {
                                addEXP()
                                showFeedback = true
                                showCorrectOverlay = true
                                playSound(success: true)
                                DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
                                    showCorrectOverlay = false
                                    showFeedback = false
                                    generateQuestion()
                                }
                            } else {
                                showFeedback = true
                                showCorrectOverlay = false
                                playSound(success: false)
                                DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
                                    showFeedback = false
                                }
                            }
                        }
                    }) {
                        HStack {
                            Image(systemName: "paperplane.fill")
                            Text(appLanguage == "ms" ? "Hantar" : "Submit")
                        }
                        .font(.system(size: 16, weight: .semibold))
                        .padding(.horizontal, 20)
                        .padding(.vertical, 10)
                        .background(Color.blue)
                        .foregroundColor(.white)
                        .cornerRadius(22)
                        .shadow(color: Color.black.opacity(0.1), radius: 3, x: 0, y: 2)
                        .frame(height: 45)
                    }

                    Button(action: {
                        paths.removeAll()
                        colors.removeAll()
                        currentPath = Path()
                        isLocked = false
                        lockedVerticalLines = nil
                        generateQuestion()
                    }) {
                        HStack {
                            Image(systemName: "questionmark.circle.fill")
                            Text(appLanguage == "ms" ? "Seterusnya" : "Next")
                        }
                        .font(.system(size: 16, weight: .semibold))
                        .padding(.horizontal, 20)
                        .padding(.vertical, 10)
                        .background(Color.green)
                        .foregroundColor(.white)
                        .cornerRadius(22)
                        .shadow(color: Color.black.opacity(0.1), radius: 3, x: 0, y: 2)
                        .frame(height: 45)
                    }
                }
                .padding(.horizontal)
                .padding(.top, 10)
                
                Spacer()

                ZStack {
                        drawingCanvas(geometry: geometry)
                            .frame(maxWidth: .infinity, maxHeight: .infinity)
                    
                }
            }
        }
        .onAppear {
            generateQuestion()
            paths.removeAll()
            colors.removeAll()
            currentPath = Path()
        }
        //.frame(width: geometry.size.width, height: geometry.size.height)
        .edgesIgnoringSafeArea(.all)
        .preferredColorScheme(.light)
        //.background(Color.white)
    }
}

File: FLUTTER/sigai/lib/lower tab/Mode.dart
//
//  Mode.swift
//  SIGAI
//
//  Created by Ainal syazwan Itamta on 22/04/2025.
//

import 'package:flutter/material.dart';
import 'package:audioplayers/audioplayers.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:sigai_flutter/mainfunc/calculate.dart' show countDivisions, countIntersections;
import 'package:sigai_flutter/mainfunc/progressBar.dart' show QuestionProgressBar;
import 'package:sigai_flutter/mainfunc/questions.dart' show GeneratedQuestion, generateQuestion;


class ModePage extends StatefulWidget {
  final String appLanguage;
  const ModePage({super.key, required this.appLanguage});

  @override
  State<ModePage> createState() => _ModePageState();
}

class _ModePageState extends State<ModePage> with WidgetsBindingObserver {
  final AudioPlayer _audioPlayer = AudioPlayer();
  bool isDarkMode = false;
  late String appLanguage;
  String mode = 'multiplication';
  Color selectedColor = Colors.grey;
  bool isLocked = false;
  List<List<Offset?>> strokes = [];
  List<Color> strokeColors = [];
  int questionDigits = 1;

  int level = 1;

  GeneratedQuestion? currentQuestion;

  String currentTip = "";
  int currentTipIndex = 0;
  final List<String> tipsMs = [
    "Pilih warna ikut kod: 🟣 -100, 🟡 -10, ⚫ 0, 🟢 1, 🔵 10, 🔴 100",
    "Untuk darab, kunci tidak digunakan.",
    "Untuk bahagi, lukis garis menegak, kunci, kemudian lukis garisan melintang.",
    "Tekan 'Hantar' untuk semak jawapan!",
    "Tekan 'Seterusnya' untuk skip ke soalan seterusnya"
  ];
  final List<String> tipsEn = [
    "Pick colors based on code: 🟣 -100, 🟡 -10, ⚫ 0, 🟢 1, 🔵 10, 🔴 100",
    "For multiplication, lock is not used.",
    "For division, draw vertical lines, lock, then draw horizontal lines.",
    "Press 'Submit' to check the answer!",
    "Press 'Next' to skip to the next question"
  ];

  int correctAnswers = 0;
  int totalQuestions = 0;

  @override
  void initState() {
    super.initState();
    appLanguage = widget.appLanguage;
    _loadProgress();
    currentTip = appLanguage == 'ms' ? tipsMs[0] : tipsEn[0];
    currentQuestion = generateQuestion(mode: mode, digits: questionDigits);
    WidgetsBinding.instance.addObserver(this);
  }

  void _loadProgress() async {
    final prefs = await SharedPreferences.getInstance();
    final loadedLevel = prefs.getInt('mode_level') ?? 1;
    final loadedCorrect = prefs.getInt('mode_correctAnswers') ?? 0;
    final loadedTotal = prefs.getInt('mode_totalQuestions') ?? 0;
    final loadedMode = prefs.getString('mode_type') ?? 'multiplication';
    final loadedDigits = prefs.getInt('mode_digits') ?? 1;
    final loadedLocked = prefs.getBool('mode_isLocked') ?? false;
    setState(() {
      level = loadedLevel;
      correctAnswers = loadedCorrect;
      totalQuestions = loadedTotal;
      mode = loadedMode;
      questionDigits = loadedDigits;
      isLocked = loadedLocked;
    });
  }

  void _saveProgress() async {
    final prefs = await SharedPreferences.getInstance();
    prefs.setInt('mode_level', level);
    prefs.setInt('mode_correctAnswers', correctAnswers);
    prefs.setInt('mode_totalQuestions', totalQuestions);
    prefs.setString('mode_type', mode);
    prefs.setInt('mode_digits', questionDigits);
    prefs.setBool('mode_isLocked', isLocked);
  }

  @override
  Widget build(BuildContext context) {
    final bool dark = isDarkMode || Theme.of(context).brightness == Brightness.dark;
    final gradientColors = dark
        ? [Colors.black,  const Color.fromARGB(255, 57, 56, 56), const Color.fromARGB(255, 130, 130, 130), Colors.white,]
        : [
            const Color(0xFFEBFAFF),
            Colors.white,
            const Color(0xFFDCEFFF),
            Colors.white,
            const Color(0xFFFFDCE6),
          ];

    final int maxQuestions = 5 + (level * 3);
    final double progress = correctAnswers / maxQuestions;
    //print("PROGRESS = $progress, correct = $correctAnswers, level = $level");

    return Scaffold(
      body: Container(
        decoration: BoxDecoration(
          gradient: LinearGradient(
            colors: gradientColors,
            begin: Alignment.topCenter,
            end: Alignment.bottomCenter,
          ),
        ),
        child: SafeArea(
          child: Column(
            children: [
            SizedBox(
              width: double.infinity,
              child: QuestionProgressBar(
                correctAnswers: correctAnswers,
                totalQuestions: totalQuestions,
                level: level,
                progress: progress,
              ),
            ),
              _buildResultDisplay(),
              const SizedBox(height: 16),
              _buildTopControls(),
              const SizedBox(height: 16),
              Expanded(
                child: Column(
                  children: [
                    Expanded(child: _buildCanvasPlaceholder()),
                  ],
                ),
              ),
              _buildBottomButtons(),
            ],
          ),
        ),
      ),
    );
  }

  
  PopupMenuItem<Color> _colorOption(Color color, String label) {
    return PopupMenuItem(
      value: color,
      child: Row(
        children: [
          Icon(Icons.circle, color: color),
          const SizedBox(width: 10),
          Text(label),
        ],
      ),
    );
  }

  Widget _buildTopControls() {
    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 12),
      child: Wrap(
        spacing: 10,
        runSpacing: 10,
        alignment: WrapAlignment.center,
        children: [
          IconButton(
            icon: const Icon(Icons.tips_and_updates, size: 32, color: Colors.orange),
            onPressed: () {
              List<String> tips = appLanguage == 'ms' ? tipsMs : tipsEn;
              setState(() {
                currentTipIndex = (currentTipIndex + 1) % tips.length;
                currentTip = tips[currentTipIndex];
              });
              ScaffoldMessenger.of(context).showSnackBar(SnackBar(content: Text(currentTip)));
            },
          ),
          IconButton(
            icon: const Icon(Icons.undo, size: 32),
            color: Colors.grey,
            onPressed: () {
              setState(() {
                if (strokes.isNotEmpty) {
                  strokes.removeLast();
                  strokeColors.removeLast();
                }
              });
            },
          ),
  
          PopupMenuButton<Color>(
            icon: Icon(Icons.palette, size: 32, color: selectedColor),
            onSelected: (color) {
              setState(() {
                selectedColor = color;
              });
            },
            itemBuilder: (context) => mode == 'multiplication'
                ? [
                    _colorOption(Colors.red, "🔴 100"),
                    _colorOption(Colors.blue, "🔵 10"),
                    _colorOption(Colors.green, "🟢 1"),
                    _colorOption(Colors.grey, "⚫ 0"),
                  ]
                : [
                    _colorOption(Colors.red, "🔴 100"),
                    _colorOption(Colors.blue, "🔵 10"),
                    _colorOption(Colors.green, "🟢 1"),
                    _colorOption(Colors.grey, "⚫ 0"),
                    _colorOption(Colors.yellow, "🟡 -10"),
                    _colorOption(Colors.purple, "🟣 -100"),
                  ],
          ),
          IconButton(
            icon: Icon(mode == 'multiplication' ? Icons.close : Icons.calculate, size: 32, color: Colors.grey),
            onPressed: () {
              setState(() {
                mode = mode == 'multiplication' ? 'division' : 'multiplication';
              });
            },
          ),
          IconButton(
            icon: Icon(isLocked ? Icons.lock : Icons.lock_open, size: 32, color: Colors.grey),
            onPressed: () {
              setState(() {
                isLocked = !isLocked;
              });
            },
          ),
          IconButton(
            icon: const Icon(Icons.delete_forever, size: 32, color: Colors.grey),
            onPressed: () {
              setState(() {
                strokes.clear();
                strokeColors.clear();
                isLocked = false;
              });
            },
          ),
        ],
      ),
    );
  }

  Widget _buildCanvasPlaceholder() {
    return GestureDetector(
      onPanUpdate: (details) {
        setState(() {
          if (strokes.isEmpty || strokes.last.contains(null)) {
            strokes.add([details.localPosition]);
            strokeColors.add(selectedColor);
          } else {
            strokes.last.add(details.localPosition);
          }
        });
      },
      onPanEnd: (details) {
        setState(() {
          if (strokes.isNotEmpty) {
            strokes.last.add(null);
          }
        });
      },
      child: Container(
        margin: const EdgeInsets.all(20),
        decoration: const BoxDecoration(
          color: Colors.transparent,
        ),
        child: CustomPaint(
          painter: _ModeDrawingPainter(strokes, strokeColors),
          size: Size.infinite,
        ),
      ),
    );
  }

  int getDetectedVerticalLines(List<List<Offset?>> paths) {
    int count = 0;
    for (int i = 0; i < paths.length; i++) {
      if (strokeColors[i] == Colors.grey) continue;
      final path = paths[i];
      final first = path.firstWhere((p) => p != null, orElse: () => null);
      final last = path.lastWhere((p) => p != null, orElse: () => null);
      if (first != null && last != null && (first.dx - last.dx).abs() < 30.0) {
        count++;
      }
    }
    return count;
  }

  Widget _buildResultDisplay() {
    final int garisCount = getDetectedVerticalLines(strokes);
    final result = mode == 'division'
        ? countDivisions(strokes, strokeColors)
        : countIntersections(strokes, strokeColors).toDouble();

    final bool dark = isDarkMode || Theme.of(context).brightness == Brightness.dark;

    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 20, vertical: 2),
      child: SizedBox(
        width: double.infinity,
        child: Container(
          decoration: BoxDecoration(
            borderRadius: BorderRadius.circular(16),
            gradient: LinearGradient(
              colors: dark
                  ? [const Color(0xFF323264), const Color(0xFF505096)]
                  : [const Color(0xFFFFCCE5), const Color(0xFFD2F0FF)],
              begin: Alignment.topCenter,
              end: Alignment.bottomCenter,
            ),
            boxShadow: [BoxShadow(color: Colors.black12, blurRadius: 6)],
          ),
          padding: const EdgeInsets.all(16),
          child: Column(
            children: [
              if (currentQuestion != null)
                Text(
                  currentQuestion!.questionText,
                  style: TextStyle(
                    fontSize: 18,
                    fontWeight: FontWeight.bold,
                    color: dark ? Colors.white : Colors.black87,
                  ),
                ),
              const SizedBox(height: 6),
              Text(
                mode == 'division'
                    ? (appLanguage == "ms"
                        ? "Hasil Bahagi ($garisCount garis):"
                        : "Division Result ($garisCount line/s):")
                    : (appLanguage == "ms"
                        ? "Hasil Darab:"
                        : "Multiplication Result:"),
                style: TextStyle(
                  fontSize: 20,
                  fontWeight: FontWeight.bold,
                  color: dark ? Colors.yellow.shade200 : Colors.blue.shade300,
                ),
              ),
              const SizedBox(height: 4),
              Text(
                mode == 'division'
                    ? result.toStringAsFixed(6)
                    : result.toStringAsFixed(0),
                style: TextStyle(
                  fontSize: 20,
                  fontWeight: FontWeight.w600,
                  color: dark ? Colors.yellow.shade200 : Colors.blue.shade300,
                ),
              ),
              const SizedBox(height: 4),
              Text(
                appLanguage == "ms" ? "Lukis jawapan." : "Draw your answer.",
                style: TextStyle(
                  fontSize: 16,
                  color: dark ? Colors.white70 : Colors.grey,
                ),
                textAlign: TextAlign.center,
              ),
            ],
          ),
        ),
      ),
    );
  }


  Widget _buildBottomButtons() {
    return Padding(
      padding: const EdgeInsets.only(bottom: 20),
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
        children: [
          ElevatedButton.icon(
            onPressed: _submitAnswer,
            icon: const Icon(Icons.send),
            label: Text(appLanguage == 'ms' ? 'Hantar' : 'Submit'),
            style: ElevatedButton.styleFrom(
              backgroundColor: Colors.blue,
              foregroundColor: Colors.white,
              padding: const EdgeInsets.symmetric(horizontal: 20, vertical: 10),
              shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(22)),
            ),
          ),
          ElevatedButton.icon(
            onPressed: () {
              setState(() {
                strokes.clear();
                strokeColors.clear();
                isLocked = false;
                currentQuestion = generateQuestion(mode: mode, digits: questionDigits);
              });
              _saveProgress();
            },
            icon: const Icon(Icons.navigate_next),
            label: Text(appLanguage == 'ms' ? 'Seterusnya' : 'Next'),
            style: ElevatedButton.styleFrom(
              backgroundColor: Colors.green,
              foregroundColor: Colors.white,
              padding: const EdgeInsets.symmetric(horizontal: 20, vertical: 10),
              shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(22)),
            ),
          ),
        ],
      ),
    );
  }

  void _submitAnswer() {
    final double result = mode == 'division'
        ? countDivisions(strokes, strokeColors)
        : countIntersections(strokes, strokeColors).toDouble();

    final double expected = currentQuestion?.expectedAnswer ?? -1;
    final bool isCorrect = (result - expected).abs() < 0.001;
    totalQuestions++;
    _saveProgress();

    if (isCorrect) {
      correctAnswers++;
      _saveProgress();

      final int maxQuestions = 5 + (level * 3);

      if (correctAnswers >= maxQuestions) {
        setState(() {
          level++;
          correctAnswers = 0;
          totalQuestions = 0;
        });
        _saveProgress();

        ScaffoldMessenger.of(context).showSnackBar(
          SnackBar(
            content: Text(appLanguage == 'ms'
              ? "Tahniah! Anda naik ke Level $level"
              : "Congratulations! You’ve reached Level $level"),
            backgroundColor: Colors.orange,
            duration: const Duration(seconds: 2),
          ),
        );
      }

      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(
          content: Text(
            isCorrect
                ? (appLanguage == 'ms' ? "Betul! Soalan seterusnya..." : "Correct! Next question...")
                : (appLanguage == 'ms' ? "Salah. Cuba lagi!" : "Wrong. Try again!"),
          ),
          backgroundColor: isCorrect ? Colors.green : Colors.red,
          duration: const Duration(seconds: 2),
        ),
      );

      Future.delayed(const Duration(milliseconds: 500), () {
        setState(() {
          strokes.clear();
          strokeColors.clear();
          isLocked = false;
          currentQuestion = generateQuestion(mode: mode, digits: questionDigits);
        });
        _saveProgress();
      });
    } else {
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(
          content: Text(
            isCorrect
                ? (appLanguage == 'ms' ? "Betul! Soalan seterusnya..." : "Correct! Next question...")
                : (appLanguage == 'ms' ? "Salah. Cuba lagi!" : "Wrong. Try again!"),
          ),
          backgroundColor: isCorrect ? Colors.green : Colors.red,
          duration: const Duration(seconds: 2),
        ),
      );
    }
  }
  @override
  void dispose() {
    _saveProgress();
    WidgetsBinding.instance.removeObserver(this);
    super.dispose();
  }

  @override
  void didChangeAppLifecycleState(AppLifecycleState state) {
    if (state == AppLifecycleState.paused ||
        state == AppLifecycleState.inactive ||
        state == AppLifecycleState.detached) {
      _saveProgress();
    } else if (state == AppLifecycleState.resumed) {
      _loadProgress();
    }
  }
}

class _ModeDrawingPainter extends CustomPainter {
  final List<List<Offset?>> strokes;
  final List<Color> strokeColors;

  _ModeDrawingPainter(this.strokes, this.strokeColors);

  @override
  void paint(Canvas canvas, Size size) {
    for (int s = 0; s < strokes.length; s++) {
      final paint = Paint()
        ..color = strokeColors[s]
        ..strokeCap = StrokeCap.round
        ..strokeWidth = 5.0;

      final points = strokes[s];
      for (int i = 0; i < points.length - 1; i++) {
        if (points[i] != null && points[i + 1] != null) {
          canvas.drawLine(points[i]!, points[i + 1]!, paint);
        }
      }
    }
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) => true;
}
